lib/routes/oauth.js

```diff
@@ -76,6 +78,50 @@ module.exports = (log, config, oauthdb) => {
         return oauthdb.createAuthorizationCode(sessionToken, request.payload);
       }
     },
+    {
+      method: 'POST',
+      path: '/oauth/token',
+      options: {
+        auth: {
+          // XXX TODO: To be able to fully replace the /token route from oauth-server,
+          // this route must also be able to accept 'client_secret' as Basic Auth in header.
+          mode: 'optional',
+          strategy: 'sessionToken'
+        },
+        validate: {
+          payload: Joi.alternatives().try(
+            oauthdb.api.grantTokensFromAuthorizationCode.opts.validate.payload,
+            oauthdb.api.grantTokensFromRefreshToken.opts.validate.payload,
+            oauthdb.api.grantTokensFromAssertion.opts.validate.payload.keys({
+              assertion: Joi.forbidden()
+            })
+          )
+        },
+        response: {
+          schema: Joi.alternatives().try(
+            oauthdb.api.grantTokensFromAuthorizationCode.opts.validate.response,
+            oauthdb.api.grantTokensFromRefreshToken.opts.validate.response,
+            oauthdb.api.grantTokensFromAssertion.opts.validate.response
+          )
+        }
+      },
+      handler: async function (request) {
+        const sessionToken = request.auth.credentials;
+        switch (request.payload.grant_type) {
+          case 'authorization_code':
+            return await oauthdb.grantTokensFromAuthorizationCode(request.payload);
+          case 'refresh_token':
+            return await oauthdb.grantTokensFromRefreshToken(request.payload);
+          case 'fxa-assertion':
```

Comment:
Just realized it's a bit strange to call this grant_type `fxa-assertion` whereas the other two use camel_case.

Reply:
Since the grant_type is not an assertion when calling the auth-server, should this be `session_token`?

Reply:
This is following our convention of naming non-standard fxa things "fxa-foo" in e.g. the JWT body and so-on.  Happy to take alternative suggestions, I was considering "fxa-credentials" rather than 'fxa-assertion".

Reply:
we should probably stay `session_token`. 👍 `fxa-credentials`.  we cannot really do `fxa-password` to follow OAuth spec closer.